### DISCUSSION TITLE
How to know if program has ended?

### DISCUSSION THREAD

**Discussion by eltsu7** [charmbracelet/wish#160] [Q&A] [answered]
I'm making an application that multiple people can join into, and I need information when one client disconnects.

My usecase is pretty close to the multichat example, but I need to redraw screen for all other clients when one disconnects. Any idea how to get this information?

**Comment by caarlos0** [comment, marked as answer]
hey, you can write a middleware that does something once the session ends, or do it in the end of the session itself. 

e.g. https://github.com/charmbracelet/wish/blob/main/logging/logging.go : notice it logs before and after calling `sh(s)`.

**Reply by eltsu7** [reply]
I see. I've made my middleware with MiddlewareWithProgramHandler like this:

```
func newApp() *app {
	a := new(app)

	s, err := wish.NewServer(
		wish.WithAddress(fmt.Sprintf("%s:%d", host, port)),
		wish.WithHostKeyPath(".ssh/term_info_ed25519"),
		wish.WithMiddleware(
			bm.MiddlewareWithProgramHandler(a.ProgramHandler, termenv.ANSI256),
			logging.Middleware(),
		),
	)

	if err != nil {
		log.Error("could not start server", "error", err)
	}

	a.Server = s
	return a
}
```
...
```
func (a *app) ProgramHandler(s ssh.Session) *tea.Program {
	if _, _, active := s.Pty(); !active {
		wish.Fatalln(s, "terminal is not active")
	}

	model := LoginModel{
		app:  a,
		name: s.User(),
	}

	p := tea.NewProgram(model, tea.WithOutput(s), tea.WithInput(s), tea.WithAltScreen())
	a.progs = append(a.progs, p)

	return p
}
```

So I'm guessing `sh(s)` happens somewhere outside my code currently. Any easy way around this other than making a different middleware?

**Reply by eltsu7** [reply]
Oh wait I just realized I can just add another middleware, so I don't have to modify the current middleware. I'm new to go :). Thanks for the help.

```
func disconnectMiddleware() wish.Middleware {
	return func(sh ssh.Handler) ssh.Handler {
		return func(s ssh.Session) {
			sh(s)
			fmt.Println("disconnected!")
		}
	}
}
```

**Reply by caarlos0** [reply]
Yes, exactly! Glad you figured out 😀
